File: datasquare.go
package rsmt2d

import (
	"errors"
	"fmt"
	"math"
	"sync"

	"golang.org/x/sync/errgroup"
)

// ErrUnevenChunks is thrown when non-nil shares are not all of equal size.
// Note: chunks is synonymous with shares.
var ErrUnevenChunks = errors.New("non-nil shares not all of equal size")

// dataSquare stores all data for an original data square (ODS) or extended
// data square (EDS). Data is duplicated in both row-major and column-major
// order in order to be able to provide zero-allocation column slices.
type dataSquare struct {
	squareRow    [][][]byte // row-major
	squareCol    [][][]byte // col-major
	dataMutex    sync.Mutex
	width        uint
	shareSize    uint
	rowRoots     [][]byte
	colRoots     [][]byte
	createTreeFn TreeConstructorFn
}

// newDataSquare populates the data square from the supplied data and treeCreator.
// No root calculation is performed.
// data may have nil values.
func newDataSquare(data [][]byte, treeCreator TreeConstructorFn, shareSize uint) (*dataSquare, error) {
	width := int(math.Ceil(math.Sqrt(float64(len(data)))))
	if width*width != len(data) {
		// TODO: export this error and modify chunks to shares
		return nil, errors.New("number of chunks must be a square number")
	}

	for _, d := range data {
		if d != nil && len(d) != int(shareSize) {
			return nil, ErrUnevenChunks
		}
	}

	squareRow := make([][][]byte, width)
	for rowIdx := 0; rowIdx < width; rowIdx++ {
		squareRow[rowIdx] = data[rowIdx*width : rowIdx*width+width]

		for colIdx := 0; colIdx < width; colIdx++ {
			if squareRow[rowIdx][colIdx] != nil && len(squareRow[rowIdx][colIdx]) != int(shareSize) {
				return nil, ErrUnevenChunks
			}
		}
	}

	squareCol := make([][][]byte, width)
	for colIdx := 0; colIdx < width; colIdx++ {
		squareCol[colIdx] = make([][]byte, width)
		for rowIdx := 0; rowIdx < width; rowIdx++ {
			squareCol[colIdx][rowIdx] = data[rowIdx*width+colIdx]
		}
	}

	return &dataSquare{
		squareRow:    squareRow,
		squareCol:    squareCol,
		width:        uint(width),
		shareSize:    shareSize,
		createTreeFn: treeCreator,
	}, nil
}

// extendSquare extends the original data square by extendedWidth and fills
// the extended quadrants with fillerShare.
func (ds *dataSquare) extendSquare(extendedWidth uint, fillerShare []byte) error {
	if uint(len(fillerShare)) != ds.shareSize {
		// TODO: export this error and rename chunk to share
		return errors.New("filler chunk size does not match data square chunk size")
	}

	newWidth := ds.width + extendedWidth
	newSquareRow := make([][][]byte, newWidth)

	fillerExtendedRow := make([][]byte, extendedWidth)
	for i := uint(0); i < extendedWidth; i++ {
		fillerExtendedRow[i] = fillerShare
	}

	fillerRow := make([][]byte, newWidth)
	for i := uint(0); i < newWidth; i++ {
		fillerRow[i] = fillerShare
	}

	row := make([][]byte, ds.width)
	for i := uint(0); i < ds.width; i++ {
		copy(row, ds.squareRow[i])
		newSquareRow[i] = append(row, fillerExtendedRow...)
	}

	for i := ds.width; i < newWidth; i++ {
		newSquareRow[i] = make([][]byte, newWidth)
		copy(newSquareRow[i], fillerRow)
	}

	ds.squareRow = newSquareRow

	newSquareCol := make([][][]byte, newWidth)
	for colIdx := uint(0); colIdx < newWidth; colIdx++ {
		newSquareCol[colIdx] = make([][]byte, newWidth)
		for rowIdx := uint(0); rowIdx < newWidth; rowIdx++ {
			newSquareCol[colIdx][rowIdx] = newSquareRow[rowIdx][colIdx]
		}
	}
	ds.squareCol = newSquareCol
	ds.width = newWidth

	ds.resetRoots()

	return nil
}

func (ds *dataSquare) rowSlice(rowIdx uint, fromIdx uint, length uint) [][]byte {
	return ds.squareRow[rowIdx][fromIdx : fromIdx+length]
}

// row returns a row slice.
// Do not modify this slice directly, instead use SetCell.
func (ds *dataSquare) row(rowIdx uint) [][]byte {
	return ds.rowSlice(rowIdx, 0, ds.width)
}

func (ds *dataSquare) setRowSlice(rowIdx uint, fromIdx uint, newRow [][]byte) error {
	for i := uint(0); i < uint(len(newRow)); i++ {
		if len(newRow[i]) != int(ds.shareSize) {
			// TODO: export this error and rename chunk to share
			return errors.New("invalid chunk size")
		}
	}
	if fromIdx+uint(len(newRow)) > ds.width {
		return fmt.Errorf("cannot set row slice at (%d, %d) of length %d: because it would exceed the data square width %d", rowIdx, fromIdx, len(newRow), ds.width)
	}

	ds.dataMutex.Lock()
	defer ds.dataMutex.Unlock()

	for i := uint(0); i < uint(len(newRow)); i++ {
		ds.squareRow[rowIdx][fromIdx+i] = newRow[i]
		ds.squareCol[fromIdx+i][rowIdx] = newRow[i]
	}

	ds.resetRoots()

	return nil
}

func (ds *dataSquare) colSlice(rowIdx uint, colIdx uint, length uint) [][]byte {
	return ds.squareCol[colIdx][rowIdx : rowIdx+length]
}

// col returns a column slice.
// Do not modify this slice directly, instead use SetCell.
func (ds *dataSquare) col(colIdx uint) [][]byte {
	return ds.colSlice(0, colIdx, ds.width)
}

func (ds *dataSquare) setColSlice(colIdx uint, fromIdx uint, newCol [][]byte) error {
	for i := uint(0); i < uint(len(newCol)); i++ {
		if len(newCol[i]) != int(ds.shareSize) {
			// TODO: export this error and rename chunk to share
			return errors.New("invalid chunk size")
		}
	}
	if fromIdx+uint(len(newCol)) > ds.width {
		return fmt.Errorf("cannot set col slice at (%d, %d) of length %d: because it would exceed the data square width %d", fromIdx, colIdx, len(newCol), ds.width)
	}

	ds.dataMutex.Lock()
	defer ds.dataMutex.Unlock()

	for i := uint(0); i < uint(len(newCol)); i++ {
		ds.squareRow[fromIdx+i][colIdx] = newCol[i]
		ds.squareCol[colIdx][fromIdx+i] = newCol[i]
	}

	ds.resetRoots()

	return nil
}

func (ds *dataSquare) resetRoots() {
	// don't write nil if it's already nil
	// this prevents rewriting nil into shared memory slot
	// when resetRoots is used from multiple routines
	if ds.rowRoots != nil {
		ds.rowRoots = nil
	}
	if ds.colRoots != nil {
		ds.colRoots = nil
	}
}

func (ds *dataSquare) computeRoots() error {
	var g errgroup.Group

	rowRoots := make([][]byte, ds.width)
	colRoots := make([][]byte, ds.width)

	for i := uint(0); i < ds.width; i++ {
		i := i // https://go.dev/doc/faq#closures_and_goroutines
		g.Go(func() error {
			rowRoot, err := ds.getRowRoot(i)
			if err != nil {
				return err
			}
			rowRoots[i] = rowRoot
			return nil
		})

		g.Go(func() error {
			colRoot, err := ds.getColRoot(i)
			if err != nil {
				return err
			}
			colRoots[i] = colRoot
			return nil
		})
	}

	err := g.Wait()
	if err != nil {
		return err
	}

	ds.rowRoots = rowRoots
	ds.colRoots = colRoots
	return nil
}

// getRowRoots returns the Merkle roots of all the rows in the square.
func (ds *dataSquare) getRowRoots() ([][]byte, error) {
	if ds.rowRoots == nil {
		err := ds.computeRoots()
		if err != nil {
			return nil, err
		}
	}

	return ds.rowRoots, nil
}

// getRowRoot calculates and returns the root of the selected row. Note: unlike
// the getRowRoots method, getRowRoot does not write to the built-in cache.
// Returns an error if the row is incomplete (i.e. some shares are nil).
func (ds *dataSquare) getRowRoot(rowIdx uint) ([]byte, error) {
	if ds.rowRoots != nil {
		return ds.rowRoots[rowIdx], nil
	}

	tree := ds.createTreeFn(Row, rowIdx)
	row := ds.row(rowIdx)
	if !isComplete(row) {
		return nil, errors.New("can not compute root of incomplete row")
	}
	for _, d := range row {
		err := tree.Push(d)
		if err != nil {
			return nil, err
		}
	}

	return tree.Root()
}

// getColRoots returns the Merkle roots of all the columns in the square.
func (ds *dataSquare) getColRoots() ([][]byte, error) {
	if ds.colRoots == nil {
		err := ds.computeRoots()
		if err != nil {
			return nil, err
		}
	}

	return ds.colRoots, nil
}

// getColRoot calculates and returns the root of the selected row. Note: unlike
// the getColRoots method, getColRoot does not write to the built-in cache.
// Returns an error if the column is incomplete (i.e. some shares are nil).
func (ds *dataSquare) getColRoot(colIdx uint) ([]byte, error) {
	if ds.colRoots != nil {
		return ds.colRoots[colIdx], nil
	}

	tree := ds.createTreeFn(Col, colIdx)
	col := ds.col(colIdx)
	if !isComplete(col) {
		return nil, errors.New("can not compute root of incomplete column")
	}
	for _, d := range col {
		err := tree.Push(d)
		if err != nil {
			return nil, err
		}
	}

	return tree.Root()
}

// GetCell returns a copy of a specific cell.
func (ds *dataSquare) GetCell(rowIdx uint, colIdx uint) []byte {
	if ds.squareRow[rowIdx][colIdx] == nil {
		return nil
	}
	cell := make([]byte, ds.shareSize)
	copy(cell, ds.squareRow[rowIdx][colIdx])
	return cell
}

// SetCell sets a specific cell. The cell to set must be `nil`. Returns an error
// if the cell to set is not `nil` or newShare is not the correct size.
func (ds *dataSquare) SetCell(rowIdx uint, colIdx uint, newShare []byte) error {
	if ds.squareRow[rowIdx][colIdx] != nil {
		return fmt.Errorf("cannot set cell (%d, %d) as it already has a value %x", rowIdx, colIdx, ds.squareRow[rowIdx][colIdx])
	}
	if len(newShare) != int(ds.shareSize) {
		// TODO: export this error and rename chunk to share
		return fmt.Errorf("cannot set cell with chunk size %d because dataSquare chunk size is %d", len(newShare), ds.shareSize)
	}
	ds.squareRow[rowIdx][colIdx] = newShare
	ds.squareCol[colIdx][rowIdx] = newShare
	ds.resetRoots()
	return nil
}

// Flattened returns the concatenated rows of the data square.
func (ds *dataSquare) Flattened() [][]byte {
	flattened := make([][]byte, 0, ds.width*ds.width)
	for _, data := range ds.squareRow {
		flattened = append(flattened, data...)
	}

	return flattened
}

// isComplete returns true if all the shares are non-nil.
func isComplete(shares [][]byte) bool {
	for _, share := range shares {
		if share == nil {
			return false
		}
	}
	return true
}

File: src/data_square.rs
//! DataSquare implementation for managing 2D data structures

use crate::{Axis, Error, Result, TreeConstructorFn};
use std::sync::Mutex;

/// DataSquare stores all data for an original data square (ODS) or extended data square (EDS).
/// Data is duplicated in both row-major and column-major order to provide zero-allocation slices.
pub struct DataSquare {
    square_row: Vec<Vec<Vec<u8>>>,  // row-major
    square_col: Vec<Vec<Vec<u8>>>,  // col-major
    data_mutex: Mutex<()>,
    width: u32,
    share_size: u32,
    row_roots: Option<Vec<Vec<u8>>>,
    col_roots: Option<Vec<Vec<u8>>>,
    create_tree_fn: TreeConstructorFn,
}

impl DataSquare {
    /// Create a new DataSquare from flattened data
    pub fn new(
        data: Vec<Vec<u8>>,
        tree_creator: TreeConstructorFn,
        share_size: u32,
    ) -> Result<Self> {
        // Validate share sizes
        for share in &data {
            if !share.is_empty() && share.len() != share_size as usize {
                return Err(Error::UnevenChunks);
            }
        }
        
        let width = (data.len() as f64).sqrt() as u32;
        if (width * width) as usize != data.len() {
            return Err(Error::InvalidChunkSize);
        }
        
        // Initialize row-major storage
        let mut square_row = vec![vec![Vec::new(); width as usize]; width as usize];
        for (i, share) in data.iter().enumerate() {
            let row = i / width as usize;
            let col = i % width as usize;
            square_row[row][col] = share.clone();
        }
        
        // Initialize column-major storage
        let mut square_col = vec![vec![Vec::new(); width as usize]; width as usize];
        for col_idx in 0..width as usize {
            for row_idx in 0..width as usize {
                square_col[col_idx][row_idx] = square_row[row_idx][col_idx].clone();
            }
        }
        
        Ok(Self {
            square_row,
            square_col,
            data_mutex: Mutex::new(()),
            width,
            share_size,
            row_roots: None,
            col_roots: None,
            create_tree_fn: tree_creator,
        })
    }
    
    /// Get a row slice
    pub fn row(&self, row_idx: u32) -> Result<&Vec<Vec<u8>>> {
        if row_idx >= self.width {
            return Err(Error::InvalidChunkSize);
        }
        Ok(&self.square_row[row_idx as usize])
    }
    
    /// Get a column slice
    pub fn col(&self, col_idx: u32) -> Result<&Vec<Vec<u8>>> {
        if col_idx >= self.width {
            return Err(Error::InvalidChunkSize);
        }
        Ok(&self.square_col[col_idx as usize])
    }
    
    /// Set a cell in the data square
    pub fn set_cell(&mut self, row_idx: u32, col_idx: u32, data: Vec<u8>) -> Result<()> {
        if row_idx >= self.width || col_idx >= self.width {
            return Err(Error::InvalidChunkSize);
        }
        
        if !data.is_empty() && data.len() != self.share_size as usize {
            return Err(Error::UnevenChunks);
        }
        
        {
            let _guard = self.data_mutex.lock().unwrap();
            
            // Update both row-major and column-major storage
            self.square_row[row_idx as usize][col_idx as usize] = data.clone();
            self.square_col[col_idx as usize][row_idx as usize] = data;
        }
        
        // Reset cached roots
        self.reset_roots();
        
        Ok(())
    }
    
    /// Get a cell from the data square
    pub fn get_cell(&self, row_idx: u32, col_idx: u32) -> Result<&Vec<u8>> {
        if row_idx >= self.width || col_idx >= self.width {
            return Err(Error::InvalidChunkSize);
        }
        Ok(&self.square_row[row_idx as usize][col_idx as usize])
    }
    
    /// Get the width of the data square
    pub fn width(&self) -> u32 {
        self.width
    }
    
    /// Reset cached roots
    fn reset_roots(&mut self) {
        self.row_roots = None;
        self.col_roots = None;
    }
    
    /// Compute row roots
    pub fn row_roots(&mut self) -> Result<Vec<Vec<u8>>> {
        if let Some(ref roots) = self.row_roots {
            return Ok(roots.clone());
        }
        
        let mut roots = Vec::new();
        for row_idx in 0..self.width {
            let root = self.compute_row_root(row_idx)?;
            roots.push(root);
        }
        
        self.row_roots = Some(roots.clone());
        Ok(roots)
    }
    
    /// Compute column roots
    pub fn col_roots(&mut self) -> Result<Vec<Vec<u8>>> {
        if let Some(ref roots) = self.col_roots {
            return Ok(roots.clone());
        }
        
        let mut roots = Vec::new();
        for col_idx in 0..self.width {
            let root = self.compute_col_root(col_idx)?;
            roots.push(root);
        }
        
        self.col_roots = Some(roots.clone());
        Ok(roots)
    }
    
    /// Compute root for a specific row
    fn compute_row_root(&self, row_idx: u32) -> Result<Vec<u8>> {
        let mut tree = (self.create_tree_fn)(Axis::Row, row_idx);
        let row = self.row(row_idx)?;
        
        for share in row {
            if share.is_empty() {
                return Err(Error::IncompleteAxis { axis: Axis::Row });
            }
            tree.push(share)?;
        }
        
        tree.root().map_err(|e| Error::Tree(e.to_string()))
    }
    
    /// Compute root for a specific column
    fn compute_col_root(&self, col_idx: u32) -> Result<Vec<u8>> {
        let mut tree = (self.create_tree_fn)(Axis::Col, col_idx);
        let col = self.col(col_idx)?;
        
        for share in col {
            if share.is_empty() {
                return Err(Error::IncompleteAxis { axis: Axis::Col });
            }
            tree.push(share)?;
        }
        
        tree.root().map_err(|e| Error::Tree(e.to_string()))
    }
    
    /// Get flattened representation of the data square
    pub fn flattened(&self) -> Vec<Vec<u8>> {
        let mut result = Vec::new();
        for row in &self.square_row {
            for share in row {
                result.push(share.clone());
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tree::new_default_tree;

    #[test]
    fn test_data_square_creation() {
        let data = vec![
            vec![1, 2], vec![3, 4],
            vec![5, 6], vec![7, 8],
        ];
        let ds = DataSquare::new(data, new_default_tree, 2);
        assert!(ds.is_ok());
        
        let ds = ds.unwrap();
        assert_eq!(ds.width(), 2);
    }
    
    #[test]
    fn test_data_square_operations() {
        let data = vec![
            vec![1, 2], vec![3, 4],
            vec![5, 6], vec![7, 8],
        ];
        let mut ds = DataSquare::new(data, new_default_tree, 2).unwrap();
        
        // Test get_cell
        assert_eq!(ds.get_cell(0, 0).unwrap(), &vec![1, 2]);
        assert_eq!(ds.get_cell(1, 1).unwrap(), &vec![7, 8]);
        
        // Test set_cell
        assert!(ds.set_cell(0, 0, vec![9, 10]).is_ok());
        assert_eq!(ds.get_cell(0, 0).unwrap(), &vec![9, 10]);
        
        // Test invalid indices
        assert!(ds.get_cell(2, 0).is_err());
        assert!(ds.set_cell(0, 2, vec![1, 2]).is_err());
    }
}